test_semver.py

```diff
@@ -655,3 +656,45 @@ def test_should_process_raise_error(capsys):
     assert rc != 0
     captured = capsys.readouterr()
     assert captured.err.startswith("ERROR")
+
+
+@pytest.mark.parametrize("version,parts,expected", [
+    ("3.4.5", dict(major=2), '2.4.5'),
+    ("3.4.5", dict(major="2"), '2.4.5'),
+    ("3.4.5", dict(major=2, minor=5), '2.5.5'),
+    ("3.4.5", dict(minor=2), '3.2.5'),
+    ("3.4.5", dict(major=2, minor=5, patch=10), '2.5.10'),
+    ("3.4.5", dict(major=2, minor=5, patch=10, prerelease="rc1"),
+     '2.5.10-rc1'),
+    ("3.4.5", dict(major=2, minor=5, patch=10, prerelease="rc1", build="b1"),
+     '2.5.10-rc1+b1'),
+    ("3.4.5-alpha.1.2", dict(major=2), '2.4.5-alpha.1.2'),
+    ("3.4.5-alpha.1.2", dict(build="x1"), '3.4.5-alpha.1.2+x1'),
+    ("3.4.5+build1", dict(major=2), '2.4.5+build1'),
+])
+def test_replace_method_replaces_requested_parts(version, parts, expected):
+    assert replace(version, **parts) == expected
+
+
+def test_should_raise_keyerror_for_invalid_key_in_replace():
+    with pytest.raises(TypeError):
```

Comment:
Why `TypeError`? Shouldn't this be a ValueError as in https://github.com/k-bx/python-semver/pull/156/files#diff-a9f8e0a40655687a794b987c4b31ff9fR699?

Reply:
Well, it's because if you pass an unknown key to `semver.VersionInfo`, the `__init__` [raises a TypeError](https://github.com/tomschr/python-semver/blob/feature/144-replace/semver.py#L309), for example:

```python
>>> import semver
>>> semver.VersionInfo(major=1, foo=42)
Traceback (most recent call last)
 ...
TypeError: __init__() got an unexpected keyword argument 'foo'
```

The current implementation catches the TypeError and improves the error message:

```python
>>> semver.VersionInfo(major=1).replace(foo=42) 
Traceback (most recent call last)
 ...
TypeError: replace() got 1 unexpected keyword argument(s): foo
```

Of course, I can rewrite the exception to KeyError or ValueError if you like. However, that means, we need to adapt that in [`semver.replace`](https://github.com/tomschr/python-semver/blob/feature/144-replace/semver.py#L729)  as well.

Reply:
My point is: The free function and the member function should do the same thing. Why does one raise a ValueError while the other raises a TypeError in the same situation? As Python itself raises a TypeError on unexpected keyword arg then the improved error should be a TypeError too. So it seems https://github.com/k-bx/python-semver/pull/156/files#diff-a9f8e0a40655687a794b987c4b31ff9fR699 need to be adapted.

I've just seen: They test different things. So it is the name that is wrong.

Reply:
Made a rename suggestion. Now it should be all good :)

Reply:
> My point is: The free function and the member function should do the same thing.

I agree. :+1: 

> So it seems https://github.com/k-bx/python-semver/pull/156/files#diff-a9f8e0a40655687a794b987c4b31ff9fR699 need to be adapted.

Hmn, ... :thinking: ... after looking at both implementations, it seems, it's easier to change the (already existing) TypeError into a ValueError in [line 315](https://github.com/tomschr/python-semver/blob/feature/144-replace/semver.py#L315). 

@Flamefire 
Would that be an option? Or should I better stay with TypeError (and adapting `semver.replace` as well)?

Thanks for all your valuable feedback. Highly appreciated! :+1:

Reply:
> after looking at both implementations, it seems, it's easier to change the (already existing) TypeError into a ValueError i

No. I noticed my mistake while typing that message, see the last sentence: One tests an invalid keyword arg which should result in a TypeError and the other tests an invalid value for a valid keyword arg which should result in a ValueError.  I missed that little difference and just assumed the functions would test the same thing.

So just adapt the name of the test function so they match the behavior (see suggestion), I think everything else is fine, so this should be the last change.

Still one possible improvement: You could test the error message too (that it contains `replace()` and the wrong keyword arg) if you wanted. I don't consider that important, but I've worked with people who do.

Reply:
> So just adapt the name of the test function so they match the behavior (see suggestion), I think everything else is fine, so this should be the last change.

Ok, thanks, I did that now.

> Still one possible improvement: You could test the error message too (that it contains replace() and the wrong keyword arg) if you wanted. I don't consider that important, but I've worked with people who do.

I thought of that too, but was a bit unsure if I should implement it. But you are right, it could be an improvement. Will do so. :wink:

Reply:
Implemented in commit 3a758a2